File: source/lib/SettingsHandler.cpp
/*
* Copyright (C) 2010 Wolfgang Koller
* 
* This file is part of GOFG Sports Computer.
* 
* GOFG Sports Computer is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
* 
* GOFG Sports Computer is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with GOFG Sports Computer.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "SettingsHandler.h"

template<>
SettingsHandler *Singleton<SettingsHandler>::mySelf = NULL;

void SettingsHandler::Set( std::string name, std::string value ) {
	if( this->settingsStore.find( name ) != this->settingsStore.end() ) {
		this->settingsStore[name] = value;
	}
}

void SettingsHandler::Set( std::string name, int value ) {
	char myBuf[11];	//11 chars are enough for a 32bit int
	sprintf( myBuf, "%d", value );	// Convert into to string

	// Call actual set function
	this->Set( name, (std::string) myBuf );
}


/**
 * <summary>	Sets a given setting to a boolean value </summary>
 *
 * <remarks>	Wkoller, 23.03.2011. </remarks>
 *
 * <param name="name"> 	Name of setting. </param>
 * <param name="value">	Value of setting. </param>
 */
void SettingsHandler::Set( std::string name, bool value ) {
	if( value ) {
		this->Set( name, 1 );
	}
	else {
		this->Set( name, 0 );
	}
}

int SettingsHandler::GetInt( std::string name ) {
	if( this->settingsStore.find( name ) != this->settingsStore.end() ) {
		return strtol( this->settingsStore[name].c_str(), NULL, 10 );
	}

	return -1;
}

double SettingsHandler::GetDouble( std::string name ) {
	if( this->settingsStore.find( name ) != this->settingsStore.end() ) {
		return strtod( this->settingsStore[name].c_str(), NULL );
	}

	return -1.0;
}

std::string SettingsHandler::GetString( std::string name ) {
	if( this->settingsStore.find( name ) != this->settingsStore.end() ) {
		return this->settingsStore[name];
	}

	return NULL;
}


/**
 * <summary>	Gets a boolean setting. </summary>
 *
 * <remarks>	Wkoller, 23.03.2011. </remarks>
 *
 * <param name="name">	The name of the setting. </param>
 *
 * <returns>	value of setting. </returns>
 */
bool SettingsHandler::GetBool( std::string name ) {
	if( this->settingsStore.find( name ) != this->settingsStore.end() ) {
		if( this->GetInt( name ) ) return true;
	}

	return false;
}

bool SettingsHandler::Load() {
	TiXmlDocument settingsDoc( "gsc_settings.xml" );

	// Check if settings file exists
	if( settingsDoc.LoadFile() ) {
		// Get the top-level node
		TiXmlElement *topNode = settingsDoc.FirstChildElement();

		// Cycle through our simple settings file and load all entries
		for( TiXmlElement *currElement = topNode->FirstChildElement(); currElement != NULL; currElement = currElement->NextSiblingElement() ) {
			if( this->settingsStore.find( currElement->Value() ) != this->settingsStore.end() ) {
				this->settingsStore[currElement->Value()] = currElement->GetText();
			}
		}

		return true;
	}

	return false;
}

bool SettingsHandler::Save() {
	TiXmlDocument settingsDoc;

	// Create top node of settings handler
	TiXmlElement *topNode = new TiXmlElement( "gsc_settings" );
	settingsDoc.LinkEndChild( topNode );

	// Iterate through the internal settingsstore and save all entries as xml-tags
	for( std::map<std::string,std::string>::iterator it = this->settingsStore.begin(); it != this->settingsStore.end(); it++ ) {
		TiXmlElement *settingsNode = new TiXmlElement( (*it).first.c_str() );
		settingsNode->LinkEndChild( new TiXmlText( (*it).second.c_str() ) );

		topNode->LinkEndChild( settingsNode );
	}

	// Write settings to disc
	return settingsDoc.SaveFile( "gsc_settings.xml" );
}

SettingsHandler::SettingsHandler() {
	// Define default settings here
	this->settingsStore["TrackFolder"] = "/";
	this->settingsStore["ExportFolder"] = "/";
	this->settingsStore["DefaultExportType"] = "0";
	this->settingsStore["MinLocationAccuracy"] = "25";
	this->settingsStore["SelectFolderPath"] = "/";
	this->settingsStore["WaitForGPSFix"] = "1";
	this->settingsStore["UseZephyrHxM"] = "0";
	this->settingsStore["ZephyrHxMDevice"] = "COM1";

	this->Load();
}

File: assets/www/js/settingshandler.js
/*
 * Copyright (C) 2011 Wolfgang Koller
 * 
 * This file is part of GOFG Sports Computer - http://www.gofg.at/.
 * 
 * GOFG Sports Computer is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * GOFG Sports Computer is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with GOFG Sports Computer.  If not, see <http://www.gnu.org/licenses/>.
 */

var SettingsHandler = {
	m_appDirectoryEntry : null,
	m_settingsStore : {
		"minimumaccuracy" : 25,
		"minimumaltitudeaccuracy" : 10,
		"minimumaltitudechange" : 3,
		"showdidyouknow" : 'show',
	},
	m_settingsFileEntry : null,
	
	init : function( p_appDirectoryEntry ) {
		SettingsHandler.m_appDirectoryEntry = p_appDirectoryEntry;
		
		SettingsHandler.m_appDirectoryEntry.getFile( "gsc_settings.xml", { create: true, exclusive: false }, SettingsHandler._settingsFileEntry, SettingsHandler._fileError );
	},
	
	get : function( p_key ) {
		if( p_key in SettingsHandler.m_settingsStore ) {
			return SettingsHandler.m_settingsStore[p_key];
		}
		
		return null;
	},
	
	set : function( p_key, p_value ) {
		if( p_key in SettingsHandler.m_settingsStore ) {
			SettingsHandler.m_settingsStore[p_key] = p_value;
		}
	},
	
	_save : function() {
		SettingsHandler.m_settingsFileEntry.createWriter( function( p_fileWriter ) {
			// Create settings header
			var settingsString = "<?xml version='1.0' encoding='UTF-8' ?>\n";
			settingsString += "<settings>\n";
			
			$.each( SettingsHandler.m_settingsStore, function( p_key, p_value ) {
				settingsString += "\t<" + p_key + ">" + p_value + "</" + p_key + ">\n";
			} );
			
			settingsString += '</settings>';
			
			p_fileWriter.write( settingsString );
		},
		SettingsHandler._fileError );
	},
	
	_settingsFileEntry : function( p_fileEntry ) {
		SettingsHandler.m_settingsFileEntry = p_fileEntry;

		SettingsHandler.m_settingsFileEntry.file( function( p_file ) {
			var reader = new FileReader();
			reader.onload = function( p_evt ) {
//				console.log( "Settings-File loaded" );
//				console.log( p_evt.target.result );
				
				var xmlDoc = $.parseXML( p_evt.target.result );
				
				$(xmlDoc).find( "settings > *" ).each( function(p_index, p_element) {
//					console.log( "Found setting '" + this.nodeName.toLowerCase() + "' with value '" + $(this).text() + "'" );
					
					SettingsHandler.m_settingsStore[this.nodeName.toLowerCase()] = $(this).text();
				});
			}
			reader.readAsText( p_file );
		}, this._fileError );
	},
	
	_fileError : function( p_fileError ) {
		console.log( "Error while handling settings file: " + p_fileError );
	}
};


/*
var SettingsHandler = function( p_appDirectoryEntry ) {
	this.m_appDirectoryEntry = p_appDirectoryEntry;
	this.m_settingsRoot = $( '<div></div>' );
	
	this.m_appDirectoryEntry.getFile( "gsc_settings.xml", { create: true, exclusive: false }, this._fileEntry, this._fileError );
}

SettingsHandler.prototype.setSetting = function( p_key, p_value ) {
}

SettingsHandler.prototype.getSetting = function( p_key ) {
}

SettingsHandler.prototype._fileEntry = function( p_fileEntry ) {
	this.m_settingsFileEntry = p_fileEntry;
	
	this.m_settingsFileEntry.file( function( p_file ) {
		var reader = new FileReader();
		reader.onload = function( p_evt ) {
			console.log( "Settings-File loaded" );
			console.log( m_settingsRoot );
			console.log( p_evt.target.result );
			
			$( m_settingsRoot ).append( p_evt.target.result );
		}
		reader.readAsText( p_file );
	}, this._fileError );
}

SettingsHandler.prototype._fileError = function( p_fileError ) {
	console.log( "Error while getting settings file: " + p_fileError );
}
*/